File: strings_and_arrays/questions/question_1.7/problem_7.rb
class Row_col_if_zero
  attr_reader :row_flags, :col_flags

  def initialize options={}
    @initialization_options = options
    @array = options[:array] || generate_array
    @row_flags = Array.new(@array.length, false)
    @col_flags = Array.new(@array.length, false)
  end

  def zero_out
    update_flags
    zero_out_rows_and_cols
  end

  def display_matrix
    for row in @array
      string_arr = []
      for element in row
        string_arr << element
      end
      p string_arr.join('')
    end
  end

  private
  def zero_out_rows_and_cols
    (0...@array.length).each do |index|
      (zero_out_row index) if row_flags[index]
      (zero_out_col index) if col_flags[index]
    end
  end

  def zero_out_row row
    @array[row].map! {0}
  end

  def zero_out_col col
    (0...@array.length).each do |index|
      @array[index][col] = 0
    end
  end

  def update_flags
    @array.each_with_index do |row, row_index|
      next if @row_flags[row_index]
      row.each_with_index do |element, col_index|
        next if @col_flags[col_index]
        if element == 0
          @row_flags[row_index] = true
          @col_flags[col_index] = true
          next
        end
      end
    end
  end

  def generate_array
    size = @initialization_options[:size] || 5
    create_matrix size
  end

  def create_matrix size
    arr = []
    (1..size).each do |row|
      arr_row = []
      (1..size).each do |col|
        arr_row << rand(0..9)
      end
      arr << arr_row
    end
    arr
  end



end

x = Row_col_if_zero.new
x.display_matrix
x.zero_out
p ''
p x.row_flags
p x.col_flags
x.display_matrix

File: strings_and_arrays/questions/question_1.7/problem_7.js
var generateMatrix = function(size){
  var matrix = []
  for (var i = 0; i < size; i++){
    var row = []
    for (var j = 0; j < size; j++){
      row.push(  Math.floor(  (Math.random()*10)  )  )
    };
    matrix.push(row)
  };
  return matrix
};

var displayMatrix = function(matrix){
  for (row in matrix){
    console.log(mat[row].join(''))
  }
}

var zeroOut = function(matrix){
  var length = matrix.length
  var rowFlags = []
  var colFlags = []
  for (var i=0; i<length; i++){ rowFlags.push(false); colFlags.push(false) }


  setRowAndColFlags(matrix, rowFlags, colFlags)
  processRowAndColFlags(matrix, rowFlags, colFlags)

  console.log('rowFlags: ' + rowFlags)
  console.log('colFlags: ' + colFlags)

}

var setRowAndColFlags = function(arr, rowFlags, colFlags){
  var length = arr.length
  for (var row_index = 0; row_index<length; row_index++){
    for (var col_index = 0; col_index<length; col_index++){
      if (arr[row_index][col_index] == 0){
        rowFlags[row_index] = true
        colFlags[col_index] = true
      }
    }
  }
}

var processRowAndColFlags = function(matrix, rowFlags, colFlags){
  for (var index = 0; index<rowFlags.length; index++){
    if (rowFlags[index]){
      transformRow(matrix, index)
    }
    if (colFlags[index]){
      transformCol(matrix, index)
    }
  }
}

var transformRow = function(matrix, row){
  for (var col = 0; col<matrix.length; col++){
    matrix[row][col] = 0
  }
}

var transformCol = function(matrix, col){
  for (var index = 0; index<matrix.length; index++){
    matrix[index][col] = 0;
  }
}


var mat = generateMatrix(8)
displayMatrix(mat)
console.log('XXXXXXXXXXXXXXXX')
zeroOut(mat)
displayMatrix(mat)
